std/assembly/function.ts

```diff
@@ -0,0 +1,38 @@
+type auto = i32;
+
+@final export abstract class Function<T> {
+  private _index: u32;
+  private _env: usize;
+
+  // @ts-ignore: this on getter
+  get index(this: T): u32 {
+    return load<u32>(changetype<usize>(this), offsetof<Function<T>>("_index"));
+  }
+
+  // @ts-ignore: this on getter
+  get name(this: T): string {
+    return "";
```

Comment:
is it generated in compile time or just not yet implemented?

Reply:
Perhaps will be gread extend `nameof<T>()` to `nameof<T>(value: T)` and do something like this `return nameof(this)`

Reply:
There is a difference to JS here in that a Function class represents unique signatures. There can be multiple functions with the same signature, and signatures don't have names. So far, the implementation here does what JS does for anonymous functions, essentially dropping any names that would have to be baked into the binary otherwise.

Reply:
There might be a way to make `.name` a builtin if that's important, but I assume that it turns out that we need to ship a full symbol map (names of all the functions potentially called indirectly) with each binary to achieve this. Imagine for example that there is a variable one can assign multiple functions to, then we'll have to look up the name at runtime somehow, for example by mapping from indexes to names. Not sure if the extra bytes to ship are worth it.

Reply:
Yeah. it definitely make sense do only if we could deduce this name in compile time and only if used request this like:
```ts
const fn = (a, b, c) => {}
const name  = fn.name;  // could statically reflect as "fn"

function foo(fn: Function<(a, b, c) => {}>) {
  const name = fn.name // name is unknown. So output `"anonymous"`
}
```